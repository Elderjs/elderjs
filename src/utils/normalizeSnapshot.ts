import windowsPathFix from './windowsPathFix';

const normalizeSnapshot = (val) => {
  if (Object.prototype.toString.call(val) === '[object String]') {
    return windowsPathFix(val);
  }
  if (Object.prototype.toString.call(val) === '[object Array]') {
    return val.map(normalizeSnapshot);
  }
  if (Object.prototype.toString.call(val) === '[object Object]') {
    return Object.keys(val).reduce((out, cv) => {
      // eslint-disable-next-line no-param-reassign
      out[cv] = normalizeSnapshot(val[cv]);
      return out;
    }, {});
  }
  return val;
};

export default normalizeSnapshot;
