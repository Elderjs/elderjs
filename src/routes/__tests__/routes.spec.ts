import { sep } from 'path';
import normalizeSnapshot from '../../utils/normalizeSnapshot';

process.cwd = () => 'test';

jest.mock('../../utils/svelteComponent', () => (component) => `<div class="svelteComponent">${component}</div>`);

describe('#routes', () => {
  const settings = {
    debug: {
      automagic: true,
    },
    siteUrl: '',
    distDir: 'test/public',
    rootDir: 'test',
    srcDir: 'test/src',
    hooks: {},
    $$internal: {
      clientComponents: 'test/public/svelte',
      ssrComponents: 'test/___ELDER___/compiled',
    },
  };

  it('Sets a default permalink function when undefined.', () => {
    beforeEach(() => {
      jest.resetModules();
    });
    jest.mock('glob', () => ({
      sync: jest
        .fn()
        .mockImplementationOnce(() => [
          `test${sep}src${sep}routes${sep}content${sep}route.js`,
          `test${sep}src${sep}routes${sep}content${sep}Default.svelte`,
          `test${sep}src${sep}routes${sep}home${sep}Home.svelte`,
          `test${sep}src${sep}routes${sep}home${sep}route.js`,
          `test${sep}src${sep}routes${sep}content${sep}data.js`,
          `test${sep}src${sep}routes${sep}content${sep}Layout.svelte`,
        ])
        .mockImplementationOnce(() => [
          `test${sep}___ELDER___${sep}compiled${sep}routes${sep}home${sep}Home.js`,
          `test${sep}___ELDER___${sep}compiled${sep}components${sep}AutoComplete.js`,
          `test${sep}___ELDER___${sep}compiled${sep}routes${sep}content${sep}Default.js`,
          `test${sep}___ELDER___${sep}compiled${sep}routes${sep}content${sep}Content.js`,
        ]),
    }));
    jest.mock(
      `test/src/routes/content/route.js`,
      () => ({
        all: [{ slug: 'content' }],
        template: 'Default',
        layout: 'Layout.svelte',
      }),
      { virtual: true },
    );
    jest.mock(`test/src/routes/content/data.js`, () => ({ default: { foo: 'bar' } }), { virtual: true });
    jest.mock(
      `test/src/routes/home/route.js`,
      () => ({
        all: () => [{ slug: 'home' }],
      }),
      { virtual: true },
    );

    jest.mock(`test/src/routes/content/data.js`, () => ({ default: { foo: 'bar' } }), { virtual: true });

    // eslint-disable-next-line global-require
    const routes = require('../routes').default;
    // @ts-ignore
    const routesObject = routes(settings);

    expect(routesObject.content.permalink({ request: { slug: 'content' } })).toEqual(`/content/`);
    expect(routesObject.content.permalink({ request: { slug: '/' } })).toEqual(`/`);
    expect(normalizeSnapshot(routesObject)).toMatchSnapshot();
  });

  it('Sets a single request object for all array when when no all function', () => {
    beforeEach(() => {
      jest.resetModules();
    });
    jest.mock('glob', () => ({
      sync: jest
        .fn()
        .mockImplementationOnce(() => [
          `test${sep}src${sep}routes${sep}content${sep}route.js`,
          `test${sep}src${sep}routes${sep}content${sep}Default.svelte`,
          `test${sep}src${sep}routes${sep}home${sep}Home.svelte`,
          `test${sep}src${sep}routes${sep}home${sep}route.js`,
          `test${sep}src${sep}routes${sep}SomethingCamel${sep}SomethingCamel.svelte`,
          `test${sep}src${sep}routes${sep}SomethingCamel${sep}route.js`,
          `test${sep}src${sep}routes${sep}content${sep}data.js`,
          `test${sep}src${sep}routes${sep}content${sep}Layout.svelte`,
        ])
        .mockImplementationOnce(() => [
          `test${sep}___ELDER___${sep}compiled${sep}routes${sep}home${sep}Home.js`,
          `test${sep}___ELDER___${sep}compiled${sep}components${sep}AutoComplete.js`,
          `test${sep}___ELDER___${sep}compiled${sep}routes${sep}content${sep}Default.js`,
          `test${sep}___ELDER___${sep}compiled${sep}routes${sep}content${sep}Content.js`,
          `test${sep}___ELDER___${sep}compiled${sep}SomethingCamel.js`,
        ]),
    }));
    jest.mock(
      `test/src/routes/content/route.js`,
      () => ({
        permalink: () => 'content-permalink',
        template: 'Default',
        layout: 'Layout.svelte',
      }),
      { virtual: true },
    );
    jest.mock('test/src/routes/content/data.js', () => ({ default: { foo: 'bar' } }), { virtual: true });
    jest.mock(
      'test/src/routes/home/route.js',
      () => ({
        permalink: () => 'home-permalink',
      }),
      { virtual: true },
    );
    jest.mock('test/src/routes/home/data.js', () => ({ default: { foo: 'bar' } }), { virtual: true });

    jest.mock(
      'test/src/routes/SomethingCamel/route.js',
      () => ({
        permalink: () => '/something-camel/',
      }),
      { virtual: true },
    );
    jest.mock('test/src/routes/SomethingCamel/data.js', () => ({ default: { foo: 'bar' } }), { virtual: true });

    // eslint-disable-next-line global-require
    const routes = require('../routes').default;
    // @ts-ignore
    const routesObject = routes(settings);

    expect(routesObject.content.all).toEqual([{ slug: 'content' }]);
    expect(routesObject.SomethingCamel.all).toEqual([{ slug: 'something-camel' }]);
    expect(routesObject.home.all).toEqual([{ slug: '/' }]);
    expect(normalizeSnapshot(routesObject)).toMatchSnapshot();
  });

  it('works where things are set', () => {
    jest.mock('glob', () => ({
      sync: jest
        .fn()
        .mockImplementationOnce(() => [
          `test${sep}src${sep}routes${sep}content${sep}route.js`,
          `test${sep}src${sep}routes${sep}content${sep}Default.svelte`,
          `test${sep}src${sep}routes${sep}home${sep}Home.svelte`,
          `test${sep}src${sep}routes${sep}home${sep}route.js`,
          `test${sep}src${sep}routes${sep}content${sep}data.js`,
          `test${sep}src${sep}routes${sep}content${sep}Layout.svelte`,
        ])
        .mockImplementationOnce(() => [
          `test${sep}___ELDER___${sep}compiled${sep}routes${sep}home${sep}Home.js`,
          `test${sep}___ELDER___${sep}compiled${sep}components${sep}AutoComplete.js`,
          `test${sep}___ELDER___${sep}compiled${sep}routes${sep}content${sep}Default.js`,
          `test${sep}___ELDER___${sep}compiled${sep}routes${sep}content${sep}Content.js`,
        ]),
    }));

    jest.mock(
      'test/src/routes/content/route.js',
      () => ({
        permalink: () => 'content-permalink',
        all: () => null,
        template: 'Default',
        layout: 'Layout.svelte',
      }),
      { virtual: true },
    );

    jest.mock(
      'test/src/routes/home/route.js',
      () => ({
        permalink: () => 'home-permalink',
        all: () => null,
      }),
      { virtual: true },
    );

    jest.mock('test/src/routes/content/data.js', () => ({ foo: 'bar' }), { virtual: true });
    // eslint-disable-next-line global-require
    const routes = require('../routes').default;
    // @ts-ignore
    expect(normalizeSnapshot(routes(settings))).toMatchSnapshot();
  });

  it('no template, ts imports', () => {
    jest.mock('glob', () => ({
      sync: jest
        .fn()
        .mockImplementationOnce(() => [
          `test${sep}src${sep}routes${sep}content${sep}route.js`,
          `test${sep}src${sep}routes${sep}content${sep}Default.svelte`,
          `test${sep}src${sep}routes${sep}home${sep}Home.svelte`,
          `test${sep}src${sep}routes${sep}home${sep}route.js`,
          `test${sep}src${sep}routes${sep}content${sep}data.js`,
          `test${sep}src${sep}routes${sep}content${sep}Layout.svelte`,
        ])
        .mockImplementationOnce(() => [
          `test${sep}___ELDER___${sep}compiled${sep}routes${sep}Home${sep}Home.js`,
          `test${sep}___ELDER___${sep}compiled${sep}components${sep}AutoComplete.js`,
          `test${sep}___ELDER___${sep}compiled${sep}routes${sep}content${sep}Default.js`,
          `test${sep}___ELDER___${sep}compiled${sep}routes${sep}Content${sep}Content.js`,
        ]),
    }));
    jest.mock(
      'test/src/routes/content/route.js',
      () => ({
        default: {
          permalink: () => 'content-permalink',
          all: () => null,
          layout: 'Layout.svelte',
        },
      }),
      { virtual: true },
    );

    jest.mock(
      'test/src/routes/home/route.js',
      () => ({
        default: {
          permalink: () => 'home-permalink',
          all: () => null,
        },
      }),
      { virtual: true },
    );

    jest.mock('test/src/routes/content/data.js', () => ({ default: { foo: 'bar' } }), { virtual: true });

    // eslint-disable-next-line global-require
    const routes = require('../routes').default;
    // @ts-ignore
    expect(normalizeSnapshot(routes({ ...settings, typescript: true }))).toMatchSnapshot();
  });
});
