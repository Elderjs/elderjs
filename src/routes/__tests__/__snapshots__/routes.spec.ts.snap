// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#routes Sets a default permalink function when undefined. 1`] = `
Object {
  "content": Object {
    "all": Array [
      Object {
        "slug": "content",
      },
    ],
    "data": Object {
      "foo": "bar",
    },
    "layout": "Layout",
    "layoutComponent": "<div class=\\"svelteComponent\\">Layout</div>",
    "permalink": [Function],
    "template": "Default",
    "templateComponent": "<div class=\\"svelteComponent\\">Default</div>",
  },
  "home": Object {
    "all": [Function],
    "data": [Function],
    "layout": "Layout.svelte",
    "layoutComponent": "<div class=\\"svelteComponent\\">Layout.svelte</div>",
    "permalink": [Function],
    "template": "Home.svelte",
    "templateComponent": "<div class=\\"svelteComponent\\">test/src/routes/home/Home.svelte</div>",
  },
}
`;

exports[`#routes Sets a single request object for all array when when no all function 1`] = `
Object {
  "SomethingCamel": Object {
    "all": Array [
      Object {
        "slug": "something-camel",
      },
    ],
    "data": [Function],
    "layout": "Layout.svelte",
    "layoutComponent": "<div class=\\"svelteComponent\\">Layout.svelte</div>",
    "permalink": [Function],
    "template": "SomethingCamel.svelte",
    "templateComponent": "<div class=\\"svelteComponent\\">test/src/routes/SomethingCamel/SomethingCamel.svelte</div>",
  },
  "content": Object {
    "all": Array [
      Object {
        "slug": "content",
      },
    ],
    "data": Object {
      "foo": "bar",
    },
    "layout": "Layout",
    "layoutComponent": "<div class=\\"svelteComponent\\">Layout</div>",
    "permalink": [Function],
    "template": "Default",
    "templateComponent": "<div class=\\"svelteComponent\\">Default</div>",
  },
  "home": Object {
    "all": Array [
      Object {
        "slug": "/",
      },
    ],
    "data": [Function],
    "layout": "Layout.svelte",
    "layoutComponent": "<div class=\\"svelteComponent\\">Layout.svelte</div>",
    "permalink": [Function],
    "template": "Home.svelte",
    "templateComponent": "<div class=\\"svelteComponent\\">test/src/routes/home/Home.svelte</div>",
  },
}
`;

exports[`#routes no template, ts imports 1`] = `
Object {
  "content": Object {
    "all": [Function],
    "data": Object {
      "foo": "bar",
    },
    "layout": "Layout",
    "layoutComponent": "<div class=\\"svelteComponent\\">Layout</div>",
    "permalink": [Function],
    "template": "Content.svelte",
    "templateComponent": "<div class=\\"svelteComponent\\">Content</div>",
  },
  "home": Object {
    "all": [Function],
    "data": [Function],
    "layout": "Layout.svelte",
    "layoutComponent": "<div class=\\"svelteComponent\\">Layout.svelte</div>",
    "permalink": [Function],
    "template": "Home.svelte",
    "templateComponent": "<div class=\\"svelteComponent\\">test/src/routes/home/Home.svelte</div>",
  },
}
`;

exports[`#routes works where things are set 1`] = `
Object {
  "content": Object {
    "all": [Function],
    "data": Object {
      "foo": "bar",
    },
    "layout": "Layout",
    "layoutComponent": "<div class=\\"svelteComponent\\">Layout</div>",
    "permalink": [Function],
    "template": "Default",
    "templateComponent": "<div class=\\"svelteComponent\\">Default</div>",
  },
  "home": Object {
    "all": [Function],
    "data": [Function],
    "layout": "Layout.svelte",
    "layoutComponent": "<div class=\\"svelteComponent\\">Layout.svelte</div>",
    "permalink": [Function],
    "template": "Home.svelte",
    "templateComponent": "<div class=\\"svelteComponent\\">test/src/routes/home/Home.svelte</div>",
  },
}
`;
